File: pos-flutter/test/gate5a/transaction_allocation_golden_vector_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:jshpos_pos/features/promotion/domain/transaction_allocation_engine.dart';

void main() {
  test('Flutter matches every shared PRM-003 allocation vector exactly', () {
    final root = jsonDecode(
      File(
        '../contracts/t2/gate5a/test-vectors/transaction-allocation-vectors-v1.json',
      ).readAsStringSync(),
    ) as Map<String, dynamic>;
    final rawSnapshot = root['snapshot'] as Map<String, dynamic>;
    final engine = TransactionAllocationEngine();
    final snapshot = engine.freeze(
      (rawSnapshot['lines'] as List<dynamic>).map((raw) {
        final value = raw as Map<String, dynamic>;
        return TransactionSnapshotLine(
          lineId: value['lineId'] as String,
          lineNo: value['lineNo'] as int,
          skuId: int.parse(value['skuId'] as String),
          quantity: value['quantity'] as String,
          grossAmountMinor: value['grossAmountMinor'] as int,
          discountAmountMinor: value['discountAmountMinor'] as int,
          payableAmountMinor: value['payableAmountMinor'] as int,
        );
      }).toList(),
    );
    expect(snapshot.grossAmountMinor, rawSnapshot['grossAmountMinor']);
    final history = <String, PriorRefundAllocation>{};
    for (final raw in root['refunds'] as List<dynamic>) {
      final scenario = raw as Map<String, dynamic>;
      final expected = scenario['expected'] as Map<String, dynamic>;
      final result = engine.refund(
        snapshot: snapshot,
        history: history.values.toList(),
        requests: (scenario['lines'] as List<dynamic>).map((rawLine) {
          final line = rawLine as Map<String, dynamic>;
          return RefundAllocationRequestLine(
            lineId: line['lineId'] as String,
            quantity: line['quantity'] as String,
          );
        }).toList(),
      );
      expect(
        result.grossAmountMinor,
        expected['grossAmountMinor'],
        reason: scenario['id'] as String,
      );
      expect(result.recoveredDiscountMinor, expected['recoveredDiscountMinor']);
      expect(result.refundableAmountMinor, expected['refundableAmountMinor']);
      for (final line in result.lines) {
        history[line.lineId] = PriorRefundAllocation(
          lineId: line.lineId,
          quantity: line.cumulativeQuantity,
          grossAmountMinor: line.cumulativeGrossAmountMinor,
          discountAmountMinor: line.cumulativeDiscountAmountMinor,
          payableAmountMinor: line.cumulativePayableAmountMinor,
        );
      }
    }
  });

  test('Flutter rejects over-refund duplicate lines and corrupted history', () {
    const lineId = '01K5R000000000000000000001';
    final engine = TransactionAllocationEngine();
    final snapshot = engine.freeze([
      TransactionSnapshotLine(
        lineId: lineId,
        lineNo: 1,
        skuId: 101,
        quantity: '3',
        grossAmountMinor: 1000,
        discountAmountMinor: 101,
        payableAmountMinor: 899,
      ),
    ]);
    expect(
      () => engine.refund(
        snapshot: snapshot,
        history: const [],
        requests: [RefundAllocationRequestLine(lineId: lineId, quantity: '4')],
      ),
      throwsFormatException,
    );
    expect(
      () => engine.refund(
        snapshot: snapshot,
        history: const [],
        requests: [
          RefundAllocationRequestLine(lineId: lineId, quantity: '1'),
          RefundAllocationRequestLine(lineId: lineId, quantity: '1'),
        ],
      ),
      throwsFormatException,
    );
    expect(
      () => engine.refund(
        snapshot: snapshot,
        history: [
          PriorRefundAllocation(
            lineId: lineId,
            quantity: '1',
            grossAmountMinor: 333,
            discountAmountMinor: 33,
            payableAmountMinor: 300,
          ),
        ],
        requests: [RefundAllocationRequestLine(lineId: lineId, quantity: '1')],
      ),
      throwsFormatException,
    );
  });
}

File: server/ruoyi-modules/jshpos-promotion/src/test/java/com/jingshanghui/pos/promotion/domain/TransactionAllocationGoldenVectorTest.java
package com.jingshanghui.pos.promotion.domain;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.jingshanghui.pos.promotion.domain.TransactionAllocationEngine.*;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

/** Java 读取 PRM-003 Java/Dart 共用成交与退款黄金向量。 */
class TransactionAllocationGoldenVectorTest {
    @Test
    void matchesEverySharedRefundVectorExactly() throws Exception {
        Path path = Path.of("..", "..", "..", "contracts", "t2", "gate5a", "test-vectors",
            "transaction-allocation-vectors-v1.json").normalize();
        Map<String, Object> root = new ObjectMapper().readValue(Files.readString(path), new TypeReference<>() { });
        Map<String, Object> rawSnapshot = map(root.get("snapshot"));
        TransactionAllocationEngine engine = new TransactionAllocationEngine();
        Snapshot snapshot = engine.freeze(list(rawSnapshot.get("lines")).stream().map(this::map).map(value ->
            new SnapshotLine((String) value.get("lineId"), ((Number) value.get("lineNo")).intValue(),
                Long.valueOf((String) value.get("skuId")), new BigDecimal((String) value.get("quantity")),
                number(value, "grossAmountMinor"), number(value, "discountAmountMinor"),
                number(value, "payableAmountMinor"))).toList());
        assertThat(snapshot.grossAmountMinor()).isEqualTo(number(rawSnapshot, "grossAmountMinor"));
        Map<String, PriorRefund> history = new LinkedHashMap<>();
        for (Object raw : list(root.get("refunds"))) {
            Map<String, Object> scenario = map(raw); Map<String, Object> expected = map(scenario.get("expected"));
            List<RefundRequestLine> requests = list(scenario.get("lines")).stream().map(this::map).map(value ->
                new RefundRequestLine((String) value.get("lineId"), new BigDecimal((String) value.get("quantity"))))
                .toList();
            RefundResult result = engine.refund(snapshot, new ArrayList<>(history.values()), requests);
            assertThat(result.grossAmountMinor()).as((String) scenario.get("id"))
                .isEqualTo(number(expected, "grossAmountMinor"));
            assertThat(result.recoveredDiscountMinor()).isEqualTo(number(expected, "recoveredDiscountMinor"));
            assertThat(result.refundableAmountMinor()).isEqualTo(number(expected, "refundableAmountMinor"));
            result.lines().forEach(value -> history.put(value.lineId(), new PriorRefund(value.lineId(),
                value.cumulativeQuantity(), value.cumulativeGrossAmountMinor(),
                value.cumulativeDiscountAmountMinor(), value.cumulativePayableAmountMinor())));
        }
    }

    @SuppressWarnings("unchecked") private Map<String, Object> map(Object value) { return (Map<String, Object>) value; }
    @SuppressWarnings("unchecked") private List<Object> list(Object value) { return (List<Object>) value; }
    private long number(Map<String, Object> value, String key) { return ((Number) value.get(key)).longValue(); }
}
